habitat-lab/habitat/datasets/rearrange/navmesh_utils.py

```diff
@@ -10,6 +10,142 @@
 from habitat_sim.physics import VelocityControl
 
 
+def snap_point_is_occluded(
+    target: mn.Vector3,
+    snap_point: mn.Vector3,
+    height: float,
+    sim: habitat_sim.Simulator,
+    granularity: float = 0.2,
+    target_object_id: Optional[int] = None,
+) -> bool:
+    """
+    Uses raycasting to check whether a target is occluded given a navmesh snap point.
+
+    :property target: The 3D position which should be unoccluded from the snap point.
+    :property snap_point: The navmesh snap point under consideration.
+    :property height: The height of the agent. Given navmesh snap point is grounded, the maximum height from which a visibility check should indicate non-occlusion. First check starts from this height.
+    :property sim: The Simulator instance.
+    :property granularity: The distance between raycast samples. Finer granularity is more accurate, but more expensive.
+    :property target_object_id: An optional object id which should be ignored in occlusion check.
+
+    NOTE: If agent's eye height is known and only that height should be considered, provide eye height and granulatiry > height for fastest check.
+
+    :return: whether or not the target is considered occluded from the snap_point.
+    """
+    # start from the top, assuming the agent's eyes are not at the bottom.
+    cur_height = height
+    while cur_height > 0:
+        ray = habitat_sim.geo.Ray()
+        ray.origin = snap_point + mn.Vector3(0, cur_height, 0)
+        cur_height -= granularity
+        ray.direction = target - ray.origin
```

Comment:
Hi Alex,

After a test, I found that the ray pointing to target (i.e., target object) from a `snap_point` location usually has 0.98 distance value, which is close to one. I think this is because that the target object is usually on the ground/floor, and the first thing it hits could be something else. So, I propose to change this to `ray.direction =  ray.origin - target`. Theoretically, these two are the same thing if there is nothing in between, but this offers much predicted results since we are casting a ray from the target object to the ceiling

Reply:
Some visual about this proposed change. Red circle is the resulting snap point found
<img width="715" alt="Screenshot 2023-10-04 at 9 25 01 AM" src="https://github.com/facebookr
<img width="573" alt="Screenshot 2023-10-04 at 9 25 52 AM" src="https://github.com/faceb
<img width="820" alt="Screenshot 2023-10-04 at 9 26 04 AM" src="https://github.com/facebookresearch/habitat-lab/assets/55121504/c207ccc5-4b3a-46a6-89b9-92c6e193c8c5">
ookresearch/habitat-lab/assets/55121504/44ddb51d-2597-4c3a-8bf3-c97e9d3a6396">
esearch/habitat-lab/assets/55121504/9e8b4f40-5d4a-48f9-a7cd-329970d539d7">

Reply:
Agreed that the ray direction does not matter too much theoretically.

However, there is risk with starting the ray inside the target. We expect the snap_point is in open space (because snap was possible), but we don't know about target. Starting a raycast inside a convex collision shape will miss that shape. This seems like it could introduce errors and I'm not sure what the gain here would be.